text_generation/causal_lm/cpp/speculative_decoding.cpp

```diff
@@ -0,0 +1,300 @@
+// Copyright (C) 2023-2024 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+
+#include <openvino/openvino.hpp>
+#include <cmath>
+#include <random>
+
+constexpr size_t BATCH_SIZE = 1;
+
+namespace {
+std::pair<ov::Tensor, ov::Tensor> tokenize(ov::InferRequest& tokenizer, std::string&& prompt) {
+    constexpr size_t BATCH_SIZE = 1;
+    tokenizer.set_input_tensor(ov::Tensor{ov::element::string, {BATCH_SIZE}, &prompt});
+    tokenizer.infer();
+    return {tokenizer.get_tensor("input_ids"), tokenizer.get_tensor("attention_mask")};
+}
+
+std::string detokenize(ov::InferRequest& detokenizer, std::vector<int64_t>& tokens) {
+    constexpr size_t BATCH_SIZE = 1;
+    detokenizer.set_input_tensor(ov::Tensor{ov::element::i64, {BATCH_SIZE, tokens.size()}, tokens.data()});
+    detokenizer.infer();
+    return detokenizer.get_output_tensor().data<std::string>()[0];
+}
+
+// The following reasons require TextStreamer to keep a cache of previous tokens:
+// detokenizer removes starting ' '. For example detokenize(tokenize(" a")) == "a",
+// but detokenize(tokenize("prefix a")) == "prefix a"
+// 1 printable token may consist of 2 token ids: detokenize(incomplete_token_idx) == "�"
+struct TextStreamer {
+    ov::InferRequest detokenizer;
+    std::vector<int64_t> token_cache;
+    size_t print_len = 0;
+
+    void put(int64_t token) {
+        token_cache.push_back(token);
+        std::string text = detokenize(detokenizer, token_cache);
+        if (!text.empty() && '\n' == text.back()) {
+            // Flush the cache after the new line symbol
+            std::cout << std::string_view{text.data() + print_len, text.size() - print_len};
+            token_cache.clear();
+            print_len = 0;
+	    return;
+        }
+        if (text.size() >= 3 && text.compare(text.size() - 3, 3, "�") == 0) {
+            // Don't print incomplete text
+            return;
+        }
+        std::cout << std::string_view{text.data() + print_len, text.size() - print_len} << std::flush;
+        print_len = text.size();
+    }
+
+    void end() {
+        std::string text = detokenize(detokenizer, token_cache);
+        std::cout << std::string_view{text.data() + print_len, text.size() - print_len} << '\n';
+        token_cache.clear();
+        print_len = 0;
+    }
+};
+}
+
+void forward_key_values(ov::InferRequest& request) {
+    // forwards present key/values from Results to Parameters with past key/values
+    // the first inputs are: input_ids, attention_masks, position_ids, other ones are past key/values.
+    // the first output is logits, others are present key.value
+    auto num_inputs = request.get_compiled_model().inputs().size();
+    for (size_t idx = 3; idx < num_inputs; ++idx) {
+        request.set_input_tensor(idx, request.get_output_tensor(idx - 2));
+    }
+}
+
+void init_key_values(ov::InferRequest request) {
```

Comment:
do we need to set initial shapes? why are shapes in IR?

Reply:
on batch axis and seq_len in the IR shapes are -1, if we don't initialize them with concrete values we will and up having error.
```
Shape inference input shapes {{0,4,0,64},{1,4,6,64}}
Argument shapes are inconsistent;
```